synapseclient/models/dataset.py

```diff
@@ -2617,46 +2378,66 @@ async def main():
         if not any(current_item.id == item.id for current_item in self.items):
             self.items.append(EntityRef(id=item.id, version=version))
 
-    async def remove_item_async(
+    def remove_item(
         self,
         item: Union["Dataset", "EntityRef"],
     ) -> None:
         """
-        Removes a dataset from the dataset collection.
+        Removes an entity from the dataset collection. Must be a Dataset or EntityRef.
         Effect is not seen until the dataset collection is stored.
+        Unless the version is specified, all entities with the same ID will be removed.
 
         Arguments:
-            item: The Dataset or EntityRef to remove from the collection
+            item: The Dataset to remove from the collection
+
+        Returns:
+            None
 
         Raises:
-            ValueError: If the item is not a Dataset or EntityRef
+            ValueError: If the item is not a Dataset
 
         Example: Remove a Dataset from a Dataset Collection.
             &nbsp;
 
             ```python
-            import asyncio
             from synapseclient import Synapse
             from synapseclient.models import DatasetCollection, Dataset
 
             syn = Synapse()
             syn.login()
 
-            async def main():
-                my_collection = await DatasetCollection(id="syn1234").get_async()
-                await my_collection.remove_item_async(Dataset(id="syn1235"))
-                await my_collection.store_async()
+            my_collection = DatasetCollection(id="syn1234").get()
+            my_collection.remove_item(Dataset(id="syn1235", version_number=1))
+            my_collection.store()
+            ```
 
-            asyncio.run(main())
+        Example: Remove all versions of a Dataset from a Dataset Collection.
+            &nbsp;
+
+            ```python
+            from synapseclient import Synapse
+            from synapseclient.models import DatasetCollection, Dataset
+
+            syn = Synapse()
+            syn.login()
+
+            my_collection = DatasetCollection(id="syn1234").get()
+            my_collection.remove_item(Dataset(id="syn1235"))
+            my_collection.store()
             ```
         """
         if not isinstance(item, (Dataset, EntityRef)):
             raise ValueError(
                 f"item must be a Dataset or EntityRef. {item} is a {type(item)}"
             )
 
+        version = item.version if isinstance(item, EntityRef) else item.version_number
+
         self.items = [
-            current_item for current_item in self.items if current_item.id != item.id
+            current_item
+            for current_item in self.items
+            if (version and current_item != EntityRef(id=item.id, version=version))
+            or (not version and current_item.id != item.id)
         ]
```

Comment:
This is a bit much IMO to squish into list comprehension like this. Could this be split out.

I also am concerned on the correctness of the if/or check. `self.items` could be a list of `EntityRef` or `Dataset` objects, but only `EntityRef` is considered here.

If both may be put into the `items` List the type on the field should be modified from:
`items: Optional[List["Dataset"]] = field(default_factory=list, compare=False)`

or

This and `add_item` could be modified to only use `Dataset` object types

Reply:
`self.items` can only contain EntityRefs. I'll simplify the logic